CMakeLists.txt

```diff
@@ -0,0 +1,160 @@
+
+## FreeRTOS Kernel Config ##
+# Check that <FREERTOS_KERNEL_CONFIG> contains FreeRTOSConfig.h
+if(NOT DEFINED FREERTOS_KERNEL_CONFIG)
+    message(FATAL_ERROR "<FREERTOS_KERNEL_CONFIG> is undefined. Set <FREERTOS_KERNEL_CONFIG> to the path that contains FreeRTOSConfig.h.")
+else()
+    if(NOT EXISTS "${FREERTOS_KERNEL_CONFIG}/FreeRTOSConfig.h")
+        message(FATAL_ERROR "FreeRTOSConfig.h not found in ${FREERTOS_KERNEL_CONFIG}. Set <FREERTOS_KERNEL_CONFIG> to the path that contains FreeRTOSConfig.h.")
+    endif()
+endif ()
+
+## FreRTOS Kernel Port ##
+if(FREERTOS_KERNEL_USE_CUSTOM_PORT)
+    # If a custom port is being used then check that <FREERTOS_KERNEL_PORT> contains a valid directory.
+    if(NOT DEFINED FREERTOS_KERNEL_PORT)
+        message(FATAL_ERROR "<FREERTOS_KERNEL_PORT> is undefined. When using a custom port set <FREERTOS_KERNEL_PORT> to the path that contains the port source files.")
+    else()
+        if(NOT EXISTS "${FREERTOS_KERNEL_PORT}")
+            message(FATAL_ERROR "Custom port folder: ${FREERTOS_KERNEL_PORT} does not exist.")
+        endif()
+    endif()
+else()
+    # If using a default port then check the compiler ID identified by CMake and set the port folder accordingly.
+    if(${CMAKE_C_COMPILER_ID} STREQUAL "ARMClang")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/ARMClang")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "Bruce")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/BCC/16BitDOS")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "GNU")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/GCC")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "IAR")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/IAR")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "MSVC")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/MSVC-MingW")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "OpenWatcom")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/oWatcom/16BitDOS")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "SDCC")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/SDCC")
+    else()
+        message(FATAL_ERROR "The C compiler identification is ${CMAKE_C_COMPILER_ID}. No port is available for this compiler. Check your toolchain configuration or set <FREERTOS_KERNEL_USE_CUSTOM_PORT> to use a custom FreeRTOS port.")
+    endif()
+    message(VERBOSE "The C compiler identification is: ${CMAKE_C_COMPILER_ID}. Compiler port folder is: ${FREERTOS_COMPILER_PORT_DIRECTORY}")
+
+    # Search the port folder for available port configurations.
+    file(GLOB PORTS RELATIVE "${FREERTOS_COMPILER_PORT_DIRECTORY}" CONFIGURE_DEPENDS "${FREERTOS_COMPILER_PORT_DIRECTORY}/*")
```

Comment:
Generally file globbing is discouraged as it makes the [CMake regeneration process a bit trickier](https://cmake.org/cmake/help/v3.16/command/file.html#filesystem).

I realize you did this for brevity, but it may be best to move away from this.

Reply:
My understanding of the general discouragement around file globbing sources is that fails to detect new/deleted files without regeneration. From the link: `If no CMakeLists.txt file changes when a source is added or removed then the generated build system cannot know when to ask CMake to regenerate.`

I don't think that's something to be concerned about here because the files/folders in the portable directory are fairly stable. If a user changes their kernel version to one that added/removed a port, then yes, they would have to regenerate. In my opinion that's reasonable.

Reply:
`file(GLOB ..)` approach doesn't work for include paths.
For example, we have:
* `portable/GCC/ARM_CM23/non_secure` and `/portable/GCC/ARM_CM23/secure`. Same for ARM_CM33.
* `portable/ThirdParty/GCC/Xtensa_ESP32/include`
* `portable/ThirdParty/Partner-Supported-Ports/GCC/AVR_Mega0`
* `portable/ThirdParty/XCC/Xtensa`

All of paths above are not present in cmake freertos module include path, which is not good for user.